### ISSUE TITLE
How can I run the pgo operator with a custom image?

### ISSUE THREAD

**Issue by IlyaNakhaichuk** [CrunchyData/postgres-operator#1841]
I have replaced the values ​​ccp_image_prefix and pgo_image_prefix in the helm chart
to your custom image, but I ran into the fact that when replacing ccp_image_prefix --- pgo-deploy does not start, when replacing pgo_image_prefix: ----- tries to download the image from registry.developers.crunchydata.com/crunchydata/custom_container:tag

Сould you help with this?

**Comment by jkatz** [comment]
Can you please be more specific?

- Which images are customized?
- What are the customizations?
- How are they customized? Where are they derived from?
- What is the end goal?
- What is the configuration of the YAML chart?
- Are there any logs?

**Comment by IlyaNakhaichuk** [comment]
Which images are customized?
FROM  registry.developers.crunchydata.com/crunchydata/crunchy-postgres-ha:centos7-12.3-4.4.0

What are the customizations?
Add to image wal2json12

How are they customized? Where are they derived from?
By adding layers to the container.

What is the end goal?
Database replication creation.

What is the configuration of the YAML chart?
`rbac:
    # rbac.create: if false RBAC resources should be in place
    create: true
    # rbac.useClusterAdmin: creates a ClusterRoleBinding giving cluster-admin to serviceAccount.name
    useClusterAdmin: true

# serviceAccount: settings for Service Account used by the deployer
serviceAccount:
    # serviceAccount.create: Whether to create a Service Account or not
    create: true
    # serviceAccount.name: The name of the Service Account to create or use
    name: ""

# =====================
# Configuration Options
# More info for these options can be found in the docs
# https://crunchydata.github.io/postgres-operator/latest/installation/configuration/
# =====================
archive_mode: "true"
archive_timeout: "60"
#------Bucket for test env:
backrest_aws_s3_bucket: "pgo"
#------Bucket for prod env:
#backrest_aws_s3_bucket: "pgo"
backrest_aws_s3_endpoint: "s3-eu-west-1.amazonaws.com"
#------New secret key test env:
#backrest_aws_s3_key: "$AWS_KEY_TEST"
#backrest_aws_s3_secret: "$AWS_SECRET_TEST"
#------New secret key prod:
backrest_aws_s3_key: 
backrest_aws_s3_secret: 
backrest_aws_s3_region: "eu-west-1"
backrest_aws_s3_uri_style: ""
backrest_aws_s3_verify_tls: "true"
backrest_port: ""
badger: "false"
#ccp_image_prefix: "registry.developers.crunchydata.com/crunchydata"
ccp_image_prefix: "xxxxxx.dkr.ecr.eu-west-1.amazonaws.com/pgo-operator"
ccp_image_pull_secret: ""
ccp_image_pull_secret_manifest: ""
#ccp_image_tag: "centos7-12.3-4.4.0"
ccp_image_tag: "centos7-12.3-4.4.0"
create_rbac: "true"
crunchy_debug: "false"
db_name: ""
db_password_age_days: "0"
db_password_length: "24"
db_port: "5432"
db_replicas: "0"
db_user: "root"
default_instance_memory: "2Gi"
default_pgbackrest_memory: "1Gi"
default_pgbouncer_memory: "512Mi"
delete_metrics_namespace: "false"
delete_operator_namespace: "false"
delete_watched_namespaces: "false"
disable_auto_failover: "false"
disable_fsgroup: "false"
reconcile_rbac: "true"
exporterport: "9187"
metrics: "true"
namespace: "pgo"
namespace_mode: "dynamic"
pgbadgerport: "10000"
pgo_add_os_ca_store: "false"
# ------- Password test env
#pgo_admin_password: "$ADMIN_PASSWORD_TEST"
# ------- Password prod env
pgo_admin_password: 
pgo_admin_perms: "*"
pgo_admin_role_name: "pgoadmin"
pgo_admin_username: "admin"
pgo_apiserver_port: "8443"
pgo_apiserver_url: "https://postgres-operator"
pgo_client_cert_secret: "pgo.tls"
pgo_client_container_install: "true"
pgo_client_version: "4.4.0"
pgo_cluster_admin: "false"
#pgo_disable_eventing: "false"
#pgo_disable_tls: "false"
pgo_image_prefix: "registry.developers.crunchydata.com/crunchydata"
#pgo_image_prefix: "xxxxx.dkr.ecr.eu-west-1.amazonaws.com/pgo-operator"
pgo_image_pull_secret: ""
pgo_image_pull_secret_manifest: ""
pgo_image_tag: "centos7-4.4.0"
pgo_installation_name: "dbs"
pgo_noauth_routes: ""
pgo_operator_namespace: "pgo"
pgo_tls_ca_store: ""
#pgo_tls_no_verify: "false"
pod_anti_affinity: "preferred"
pod_anti_affinity_pgbackrest: ""
pod_anti_affinity_pgbouncer: ""
scheduler_timeout: "3600"
service_type: "ClusterIP"
sync_replication: "false"
backrest_storage: "micro"
backup_storage: "micro"
primary_storage: "micro"
replica_storage: "micro"
wal_storage: ""
preserve_pg_clusters: "true"

storage1_name: "micro"
storage1_access_mode: "ReadWriteOnce"
storage1_size: "25G"
storage1_type: "dynamic"
storage1_class: "vsan-single"
`

Are there any logs?
Failed to pull image "registry.developers.crunchydata.com/crunchydata/crunchy-postgres-ha:centos7-12.3-4.4.0-328321": rpc error: code = Unknown desc = Error response from daemon: manifest for registry.developers.crunchydata.com/crunchydata/crunchy-postgres-ha:centos7-12.3-4.4.0-328321 not found: manifest unknown: manifest unknown

**Comment by jkatz** [comment]
So I'll note that PostgreSQL has supported logical replication since version 10. I've used wal2json in pre-PostgreSQL 10 applications and it certainly does help if you're trying to read directly from the WAL stream. Our containers do support logical replication.

That said, looking at your chart:

```
pgo_image_prefix: "registry.developers.crunchydata.com/crunchydata"
#pgo_image_prefix: "xxxxx.dkr.ecr.eu-west-1.amazonaws.com/pgo-operator"
```

It appears that you're pulling the images from `registry.developers.crunchydata.com/crunchydata` and not your own registry.

**Comment by IlyaNakhaichuk** [comment]
When I do this, the pgo-deploy container throws an error because it wants to pull xxxxx.dkr.ecr.eu-west-1.amazonaws.com/pgo-operator/pgo-event(pgo-scheduler; pgo-apiserver): centos7-4.4.0, but they are not there, respectively. But if you edit the postgres-operator deployment by replacing the default addresses for pgo-event, pgo-scheduler, pgo-apiserver i still get error.

```
#pgo_image_prefix: "registry.developers.crunchydata.com/crunchydata"
pgo_image_prefix: "xxxxx.dkr.ecr.eu-west-1.amazonaws.com/pgo-operator"
```

How can I take advantage of your default logical replication?

**Comment by jkatz** [comment]
Per #1816 wal2json will make it into the default container in a future release.

I think the part that confuses me a bit is why the `pgo-deployer` is trying to pull the PostgreSQL images. The deployer is just responsible for setting up the Operator itself and does not do anything to load the images prefixed `ccp_*`

**Comment by IlyaNakhaichuk** [comment]
Perhaps I did not quite put it this way, but when a custom image is substituted in 2 fields ccp_image_prefix and pgo_image_prefix, the operator does not start, when a custom image is substituted in one field of the ccp_image_prefix, the operator is launched, but no error occurs when upgrading the database using the operator.


As I understand it, due to the fact that we are changing the tag for our custom image, it tries to download dependent containers for building using the same tag from the official repository, although there are no such ones there.

Failed to pull image "registry.developers.crunchydata.com/crunchydata/crunchy-collect:centos7-12.3-4.4.0-328321": rpc error: code = Unknown desc = Error response from daemon: manifest for registry.developers.crunchydata.com/crunchydata/crunchy-collect:centos7-12.3-4.4.0-328321 not found: manifest unknown: manifest unknown

**Comment by IlyaNakhaichuk** [comment]
```
bash-4.2$ pgo status --namespace test-cl
Databases:               2
Claims:                  5
Total Volume Size:       120Gi     

Database Images:
                         1	registry.developers.crunchydata.com/crunchydata/crunchy-collect:centos7-12.3-4.4.0
                         1	registry.developers.crunchydata.com/crunchydata/crunchy-postgres-ha:centos7-12.3-4.4.0
                         1	registry.developers.crunchydata.com/crunchydata/pgo-backrest-repo:centos7-4.4.0
                         5	registry.developers.crunchydata.com/crunchydata/pgo-backrest-restore:centos7-4.3.0
```


How can I change this container to a custom one? 
"registry.developers.crunchydata.com/crunchydata/crunchy-postgres-ha:centos7-12.3-4.4.0" 

When upgrading, I can't tell you --ccp-image string, I can only when create. So I updated the operator to pass ccp-image-prefix there.

**Comment by IlyaNakhaichuk** [comment]
Do you have any solutions?

**Comment by jkatz** [comment]
I'm having a hard time following what's going on. 

I think an easier answer is that we're looking at including wal2json in the upcoming 4.5.0 release.

**Comment by IlyaNakhaichuk** [comment]
When is 4.5.0 expected?

**Comment by jkatz** [comment]
We're targeting end-of-the-month, early October. We've begun [beta testing](https://github.com/CrunchyData/postgres-operator/releases/tag/v4.5.0-beta.1).

**Comment by jkatz** [comment]
wal2json will be included in the 4.5.0 release.

**Comment by IlyaNakhaichuk** [comment]
It would be nice to make such a possibility so that you can still substitute custom images when creating a database (specifically when upgrading a database through an operator). Because the default image is not enough for various purposes. Could you consider this proposal and possibly make changes for this possibility?
